compiler/hash-parser/src/gen.rs

```diff
@@ -1362,14 +2615,83 @@ where
                     }
                 }
             }
-            _ => return Ok(None),
+
+            // Map or set type
+            TokenKind::Tree(Delimiter::Brace, tree) => {
+                self.skip_token();
+
+                let gen = self.from_stream(tree, token.span);
+
+                let lhs_type = gen.parse_type()?;
+
+                match gen.peek() {
+                    // This must be a map
+                    Some(token) if token.has_atom(TokenAtom::Colon) => {
+                        gen.skip_token();
+
+                        let rhs_type = gen.parse_type()?;
+
+                        // @@CopyPasta
+                        if gen.has_token() {
+                            gen.expected_eof()?;
+                        }
```

Comment:
Why is this pattern necessary?

Reply:
Essentially assert the the inner generator has no more tokens to consume. If it does then it is an error and the parser couldn't eat the entire generator.